frontend/src/metabase/dashboard/context/context.tsx

```diff
@@ -98,332 +105,364 @@ export const DashboardContext = createContext<
-
-  useEffect(() => {
-    if (
-      !isLoadingWithoutCards &&
-      previousIsLoadingWithoutCards &&
-      !error &&
-      dashboard
-    ) {
-      onLoadWithoutCards?.(dashboard);
-      // For whatever reason, isLoading waits for all cards to be loaded but doesn't account for the
-      // fact that there might be no dashcards. So onLoad never triggers when there are no cards,
-      // so this solves that issue for now.
-      if (dashboard?.dashcards.length === 0) {
-        onLoad?.(dashboard);
+    }, [
+      fetchDashboardCardData,
+      handleError,
+      parameterValues,
+      previousDashboard,
+      previousParameterValues,
+      previousTabId,
+      selectedTabId,
+    ]);
+
+    useEffect(() => {
+      if (dashboardIdProp !== dashboardId) {
+        fetchId(dashboardIdProp);
       }
-    }
-  }, [
-    previousIsLoadingWithoutCards,
-    dashboard,
-    onLoadWithoutCards,
-    error,
-    isLoadingWithoutCards,
-    onLoad,
-  ]);
-
-  useEffect(() => {
-    if (!isLoading && previousIsLoading && !error && dashboard) {
-      onLoad?.(dashboard);
-    }
-  }, [isLoading, previousIsLoading, dashboard, onLoad, error]);
-
-  useUnmount(() => {
-    cancelFetchDashboardCardData();
-    reset();
-    closeDashboard();
-  });
-
-  const hiddenParameterSlugs = useMemo(
-    () =>
-      getTabHiddenParameterSlugs({
-        parameters,
-        dashboard,
-        selectedTabId,
-      }),
-    [dashboard, parameters, selectedTabId],
-  );
 
-  const hideParameters = !isEditing
-    ? [hide_parameters, hiddenParameterSlugs].filter(Boolean).join(",")
-    : null;
-  // For public/static dashboards, we want to make sure that we don't show action cards
-  // so we have a filter function here to remove those. We can/will also add this
-  // functionality in the SDK in the future, which is why it's a generic prop
-  const dashboardWithFilteredCards = useMemo(() => {
-    if (dashboard && isDashcardVisible) {
-      return assoc(
-        dashboard,
-        "dashcards",
-        dashboard.dashcards.filter(isDashcardVisible),
-      );
-    }
-    return dashboard;
-  }, [dashboard, isDashcardVisible]);
+      async function fetchId(idToFetch: DashboardId) {
+        try {
+          const { id, isError } = await dispatch(
+            fetchEntityId({ type: "dashboard", id: idToFetch }),
+          );
+          if (isError || id === null) {
+            handleError({
+              status: 404,
+              message: "Not found",
+              name: "Not found",
+            });
+          }
+          setDashboardId(id);
+        } catch (e) {
+          handleError(e);
+        }
 
-  const dashboardActions =
-    typeof initDashboardActions === "function"
-      ? initDashboardActions({ isEditing, downloadsEnabled })
-      : (initDashboardActions ?? null);
+        return;
+      }
+    }, [dashboardId, dashboardIdProp, dispatch, handleError]);
+
+    useImperativeHandle(ref, (): DashboardContextProviderHandle => {
```

Comment:
This allows the parent to pass a ref to this component and do `DashboardContextProviderRef.current.refetchDashboard()`